File: src_code/1003-minimum-area-rectangle-ii/solution.java
class Solution {
    public double minAreaFreeRect(int[][] points) {
        HashMap<Integer, HashSet<Integer>> map = new HashMap();
        for (int[] point : points) {
            if (!map.containsKey(point[0])) {
                map.put(point[0], new HashSet<>());
            }
            map.get(point[0]).add(point[1]);
        }

        double minArea = Double.MAX_VALUE;
        int n = points.length;

        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                int dx1 = points[j][0] - points[i][0];
                int dy1 = points[j][1] - points[i][1];

                for (int k = j + 1; k < n; k++) {
                    int dx2 = points[k][0] - points[i][0];
                    int dy2 = points[k][1] - points[i][1];

                    if (dx1 * dx2 + dy1 * dy2 != 0) continue;

                    int x = dx1 + points[k][0];
                    int y = dy1 + points[k][1];

                    if (map.get(x) != null && map.get(x).contains(y)) {
                        double area = Math.sqrt(dx1*dx1 + dy1*dy1) * Math.sqrt(dx2*dx2 + dy2*dy2);
                        if(area < minArea){
                            minArea = area;
                        }
                    }
                }
            }
        }

        return minArea == Double.MAX_VALUE ? 0 : minArea;
    }
}

File: src_code/1003-minimum-area-rectangle-ii/solution.py
import itertools
import math

class Solution(object):
    def minAreaFreeRect(self, points):
        """
        :type points: List[List[int]]
        :rtype: float
        """
        points_set = set(map(tuple, points))
        min_area = float('inf')

        for p1, p2, p3 in itertools.combinations(points, 3):
            x1, y1 = p1
            x2, y2 = p2
            x3, y3 = p3

            # Check if p1, p2, and p3 are orthogonal
            if (x2 - x1) * (x3 - x1) + (y2 - y1) * (y3 - y1) == 0:
                # Calculate the fourth point of the potential rectangle
                x4, y4 = x2 + x3 - x1, y2 + y3 - y1

                if (x4, y4) in points_set:
                    # Calculate the area of the rectangle formed by p1, p2, p3, and p4
                    area = ((x2 - x1) ** 2 + (y2 - y1) ** 2) ** 0.5 * ((x3 - x1) ** 2 + (y3 - y1) ** 2) ** 0.5
                    min_area = min(min_area, area)

        return min_area if min_area != float('inf') else 0.0
